File: packages/native-android/android/src/main/java/com/openota/runtime/BundleManager.kt
package com.openota.runtime

import android.content.Context
import java.io.File

private const val MAX_UNCONFIRMED_BOOTS = 2

/** Snapshot returned by [BundleManager.getRuntimeInfo], mirrored 1:1 onto the JS `RuntimeInfo` type. */
data class RuntimeInfo(
    val currentVersion: String?,
    val bundleVersion: String?,
    val runtimeVersion: String?,
    val manifestVersion: Int?,
    val bundlePath: String?,
    val installTimeMillis: Long?,
    val state: RuntimeState,
    val platform: String = "android",
)

/**
 * The single orchestrator for the bundle lifecycle. `BundleManager` composes [BundleStorage],
 * [BundleVerifier], [BundleInstaller] and [BundleRollback] rather than absorbing their logic — it
 * is the only class that is allowed to drive [RuntimeStateHolder] transitions, which keeps the
 * state machine's invariants ("every transition explicit, never skipped") enforceable from one
 * place instead of re-implemented at every call site.
 *
 * `runtimeVersion` is supplied by the host app (there is no way to derive "what breaking bundle
 * format generation is this native binary" automatically) and is compared against every candidate
 * bundle's manifest before it is ever staged, per the "future-proof for breaking bundle formats"
 * requirement.
 */
class BundleManager private constructor(context: Context, private val runtimeVersion: String) {
    private val storage = BundleStorage(context)
    private val verifier = BundleVerifier(storage, runtimeVersion)
    private val rollback = BundleRollback(storage)
    private val installer = BundleInstaller(storage, rollback)

    private val stateHolder = RuntimeStateHolder(storage.readRootManifest().state)

    @Volatile
    private var pendingCandidatePath: String? = null

    @Synchronized
    fun setBundlePath(path: String) {
        val resolved = storage.resolveWithinRoot(path)
        if (!resolved.exists() || !resolved.isDirectory) {
            throw BundleVerificationException("Candidate path \"$path\" is not a directory")
        }

        pendingCandidatePath = resolved.path
        stateHolder.transition(RuntimeState.DOWNLOADED)
    }

    fun getBundlePath(): String? = storage.readRootManifest().activeBundlePath

    /**
     * Runs the full `Downloaded -> Verified -> Extracted -> Installed -> Activated` chain against
     * whatever path was last registered via [setBundlePath]. On any failure the previous `current/`
     * is left untouched and the candidate directory is deleted — the app never ends up serving a
     * half-verified bundle.
     */
    @Synchronized
    fun activateBundle(): RuntimeInfo {
        val candidatePath = pendingCandidatePath
            ?: throw NotConfiguredException("setBundlePath() must be called before activateBundle()")
        val candidateDir = File(candidatePath)

        return try {
            val verified = OpenOTALogger.timed("activate.verify") {
                stateHolder.transition(RuntimeState.VERIFIED)
                verifier.verify(candidateDir)
            }

            stateHolder.transition(RuntimeState.EXTRACTED)

            val installed = OpenOTALogger.timed("activate.install") {
                stateHolder.transition(RuntimeState.INSTALLED)
                installer.promote(verified)
            }

            stateHolder.transition(RuntimeState.ACTIVATED)
            pendingCandidatePath = null

            val manifest = RuntimeManifest(
                activeVersion = installed.manifest.version,
                activeBundlePath = installed.bundleFile.path,
                runtimeVersion = installed.manifest.runtimeVersion,
                manifestVersion = installed.manifest.manifestVersion,
                installTimeMillis = System.currentTimeMillis(),
                state = RuntimeState.ACTIVATED,
                bootConfirmed = false,
                bootAttempts = 0,
            )
            storage.writeRootManifest(manifest)

            OpenOTALogger.i("Activated ${installed.manifest.version} (runtime ${installed.manifest.runtimeVersion})")
            toRuntimeInfo(manifest)
        } catch (error: OpenOTAException) {
            OpenOTALogger.e("Activation failed: ${error.message}", error)
            stateHolder.transition(RuntimeState.FAILED)
            storage.deleteRecursively(candidateDir)
            pendingCandidatePath = null
            throw error
        }
    }

    /** Restores the previous generation from `rollback/` and makes it the active bundle. */
    @Synchronized
    fun rollbackBundle(): RuntimeInfo {
        stateHolder.transition(RuntimeState.ROLLBACK)

        val restored = rollback.restore()
        stateHolder.transition(RuntimeState.ACTIVATED)

        val manifest = RuntimeManifest(
            activeVersion = restored.manifest.version,
            activeBundlePath = restored.bundleFile.path,
            runtimeVersion = restored.manifest.runtimeVersion,
            manifestVersion = restored.manifest.manifestVersion,
            installTimeMillis = System.currentTimeMillis(),
            state = RuntimeState.ACTIVATED,
            bootConfirmed = true,
            bootAttempts = 0,
        )
        storage.writeRootManifest(manifest)
        OpenOTALogger.i("Rolled back to ${restored.manifest.version}")
        return toRuntimeInfo(manifest)
    }

    /** Wipes every downloaded/installed bundle and returns the runtime to [RuntimeState.EMBEDDED]. */
    @Synchronized
    fun clearBundle() {
        storage.deleteRecursively(storage.currentDir)
        storage.deleteRecursively(storage.rollbackDir)
        storage.deleteRecursively(storage.downloadsDir)
        storage.deleteRecursively(storage.cacheDir)
        storage.currentDir.mkdirs()
        storage.rollbackDir.mkdirs()
        storage.downloadsDir.mkdirs()
        storage.cacheDir.mkdirs()

        pendingCandidatePath = null
        stateHolder.transition(RuntimeState.EMBEDDED)
        storage.writeRootManifest(RuntimeManifest.EMPTY)
    }

    fun getRuntimeInfo(): RuntimeInfo = toRuntimeInfo(storage.readRootManifest())

    fun getActiveVersion(): String? = storage.readRootManifest().activeVersion

    /**
     * Called once JS has progressed far enough to construct this native module (see
     * `OpenOTAModule.initialize`). Two activations in a row without this ever firing is treated as
     * a boot crash-loop and triggers an automatic rollback — see the class doc on [BundleLoader]
     * for why this heuristic, rather than an explicit JS "I booted fine" call, is what v1 uses.
     */
    @Synchronized
    fun confirmBoot() {
        val manifest = storage.readRootManifest()
        if (!manifest.bootConfirmed) {
            storage.writeRootManifest(manifest.copy(bootConfirmed = true, bootAttempts = 0))
        }
    }

    /** Called at the very top of [BundleLoader.getJSBundleFile], before any JS has had a chance to run. */
    @Synchronized
    fun recordBootAttempt(): RuntimeManifest {
        val manifest = storage.readRootManifest()
        if (manifest.state != RuntimeState.ACTIVATED || manifest.bootConfirmed) {
            return manifest
        }

        val attempts = manifest.bootAttempts + 1
        if (attempts >= MAX_UNCONFIRMED_BOOTS && rollback.hasRollback()) {
            OpenOTALogger.w("Detected $attempts unconfirmed boots of ${manifest.activeVersion}; rolling back")
            return try {
                rollbackBundle().let { storage.readRootManifest() }
            } catch (error: OpenOTAException) {
                OpenOTALogger.e("Automatic rollback failed", error)
                manifest
            }
        }

        val updated = manifest.copy(bootAttempts = attempts)
        storage.writeRootManifest(updated)
        return updated
    }

    private fun toRuntimeInfo(manifest: RuntimeManifest) = RuntimeInfo(
        currentVersion = manifest.activeVersion,
        bundleVersion = manifest.activeVersion,
        runtimeVersion = manifest.runtimeVersion,
        manifestVersion = manifest.manifestVersion,
        bundlePath = manifest.activeBundlePath,
        installTimeMillis = manifest.installTimeMillis,
        state = manifest.state,
    )

    companion object {
        @Volatile
        private var instance: BundleManager? = null

        fun getInstance(context: Context, runtimeVersion: String? = null): BundleManager {
            return instance ?: synchronized(this) {
                instance ?: BundleManager(
                    context,
                    runtimeVersion ?: resolveAppVersionName(context),
                ).also { instance = it }
            }
        }

        private fun resolveAppVersionName(context: Context): String {
            val appContext = context.applicationContext
            val packageInfo = appContext.packageManager.getPackageInfo(appContext.packageName, 0)
            return packageInfo.versionName ?: "0.0.0"
        }

        /** Test-only: forces the next [getInstance] call to construct a fresh manager. */
        @androidx.annotation.VisibleForTesting
        fun resetForTesting() {
            instance = null
        }
    }
}

File: packages/native-ios/ios/BundleManager.swift
import Foundation

let maxUnconfirmedBoots = 2 // Must match BundleManager.kt's MAX_UNCONFIRMED_BOOTS exactly.

/// Snapshot returned by `BundleManager.getRuntimeInfo`, mirrored 1:1 onto the JS `RuntimeInfo` type.
struct RuntimeInfo {
    let currentVersion: String?
    let bundleVersion: String?
    let runtimeVersion: String?
    let manifestVersion: Int?
    let bundlePath: String?
    let installTimeMillis: Int64?
    let state: RuntimeState
    let platform: String = "ios"
}

/// The single orchestrator for the bundle lifecycle — the Swift counterpart of `BundleManager.kt`.
/// Implemented as an `actor` (rather than a serial `DispatchQueue`) because every operation here is
/// short-lived synchronous file I/O with no callback-based APIs to bridge, so actor isolation gives
/// the same "one state-changing operation at a time" guarantee as Kotlin's `@Synchronized` with
/// none of the manual queue plumbing, and composes cleanly with `async`/`await` call sites in the
/// TurboModule shim. File I/O inside actor methods is deliberately kept synchronous+local (never
/// blocking the *main* thread — RN already dispatches TurboModule method bodies off the main
/// queue), so actor reentrancy from suspension points is not a concern here.
actor BundleManager {
    private let storage: BundleStorage
    private let verifier: BundleVerifier
    private let rollback: BundleRollback
    private let installer: BundleInstaller
    private let runtimeVersion: String

    private var stateHolder: RuntimeStateHolder
    private var pendingCandidatePath: URL?

    init(storage: BundleStorage, runtimeVersion: String) {
        self.storage = storage
        self.runtimeVersion = runtimeVersion
        self.verifier = BundleVerifier(storage: storage, appRuntimeVersion: runtimeVersion)
        self.rollback = BundleRollback(storage: storage)
        self.installer = BundleInstaller(storage: storage, rollback: rollback)
        self.stateHolder = RuntimeStateHolder(storage.readState().state)
    }

    func setBundlePath(_ path: String) throws {
        let resolved = try storage.resolveWithinRoot(path)
        guard storage.fileExists(resolved, isDirectory: true) else {
            throw OpenOTAException.verificationFailed("Candidate path \"\(path)\" is not a directory")
        }
        pendingCandidatePath = resolved
        try stateHolder.transition(to: .downloaded)
    }

    func getBundlePath() -> String? {
        storage.readState().activeBundlePath
    }

    /// Runs the full `Downloaded -> Verified -> Extracted -> Installed -> Activated` chain against
    /// whatever path was last registered via `setBundlePath`. On any failure the previous
    /// `current/` is left untouched and the candidate directory is deleted.
    func activateBundle() throws -> RuntimeInfo {
        guard let candidateDir = pendingCandidatePath else {
            throw OpenOTAException.notConfigured("setBundlePath() must be called before activateBundle()")
        }

        do {
            try stateHolder.transition(to: .verified)
            let verified = try OpenOTALogger.timed("activate.verify") { try verifier.verify(candidateDir) }

            try stateHolder.transition(to: .extracted)

            try stateHolder.transition(to: .installed)
            let installed = try OpenOTALogger.timed("activate.install") { try installer.promote(verified) }

            try stateHolder.transition(to: .activated)
            pendingCandidatePath = nil

            let manifest = RuntimeManifest(
                activeVersion: installed.manifest.version,
                activeBundlePath: installed.bundleFile.path,
                runtimeVersion: installed.manifest.runtimeVersion,
                manifestVersion: installed.manifest.manifestVersion,
                installTimeMillis: Int64(Date().timeIntervalSince1970 * 1000),
                state: .activated,
                bootConfirmed: false,
                bootAttempts: 0
            )
            try storage.writeState(manifest)

            OpenOTALogger.i("Activated \(installed.manifest.version) (runtime \(installed.manifest.runtimeVersion))")
            return toRuntimeInfo(manifest)
        } catch {
            OpenOTALogger.e("Activation failed", error: error)
            _ = try? stateHolder.transition(to: .failed)
            storage.deleteRecursively(candidateDir)
            pendingCandidatePath = nil
            throw error
        }
    }

    /// Restores the previous generation from `rollback/` and makes it the active bundle.
    @discardableResult
    func rollbackBundle() throws -> RuntimeInfo {
        try stateHolder.transition(to: .rollback)

        let restored: VerifiedBundle
        do {
            restored = try rollback.restore()
        } catch {
            _ = try? stateHolder.transition(to: .failed)
            throw error
        }
        try stateHolder.transition(to: .activated)

        let manifest = RuntimeManifest(
            activeVersion: restored.manifest.version,
            activeBundlePath: restored.bundleFile.path,
            runtimeVersion: restored.manifest.runtimeVersion,
            manifestVersion: restored.manifest.manifestVersion,
            installTimeMillis: Int64(Date().timeIntervalSince1970 * 1000),
            state: .activated,
            bootConfirmed: true,
            bootAttempts: 0
        )
        try storage.writeState(manifest)
        OpenOTALogger.i("Rolled back to \(restored.manifest.version)")
        return toRuntimeInfo(manifest)
    }

    /// Wipes every downloaded/installed bundle and returns the runtime to `.embedded`. Never
    /// removes the app's embedded `main.jsbundle` inside the app bundle itself — only the OpenOTA
    /// tree in Application Support.
    func clearBundle() throws {
        storage.deleteRecursively(storage.currentDir)
        storage.deleteRecursively(storage.rollbackDir)
        storage.deleteRecursively(storage.downloadsDir)
        storage.deleteRecursively(storage.cacheDir)
        for dir in [storage.currentDir, storage.rollbackDir, storage.downloadsDir, storage.cacheDir] {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }

        pendingCandidatePath = nil
        try stateHolder.transition(to: .embedded)
        try storage.writeState(.empty)
    }

    func getRuntimeInfo() -> RuntimeInfo {
        toRuntimeInfo(storage.readState())
    }

    func getActiveVersion() -> String? {
        storage.readState().activeVersion
    }

    /// Called once JS has progressed far enough to construct the TurboModule (mirrors
    /// `OpenOTAModule.initialize()` -> `confirmBoot()` on Android).
    func confirmBoot() {
        var manifest = storage.readState()
        if !manifest.bootConfirmed {
            manifest.bootConfirmed = true
            manifest.bootAttempts = 0
            try? storage.writeState(manifest)
        }
    }

    /// Called at the very top of bundle resolution, before any JS has had a chance to run.
    /// Repeated launches without `confirmBoot()` past `maxUnconfirmedBoots` triggers an automatic
    /// rollback to the previous OTA generation (or, if none exists, leaves state such that the
    /// resolver falls back to the embedded bundle).
    @discardableResult
    func recordBootAttempt() -> RuntimeManifest {
        var manifest = storage.readState()
        guard manifest.state == .activated, !manifest.bootConfirmed else {
            return manifest
        }

        let attempts = manifest.bootAttempts + 1
        if attempts >= maxUnconfirmedBoots, rollback.hasRollback() {
            OpenOTALogger.w("Detected \(attempts) unconfirmed boots of \(manifest.activeVersion ?? "?"); rolling back")
            do {
                _ = try rollbackBundle()
                return storage.readState()
            } catch {
                OpenOTALogger.e("Automatic rollback failed", error: error)
                return manifest
            }
        }

        manifest.bootAttempts = attempts
        try? storage.writeState(manifest)
        return manifest
    }

    private func toRuntimeInfo(_ manifest: RuntimeManifest) -> RuntimeInfo {
        RuntimeInfo(
            currentVersion: manifest.activeVersion,
            bundleVersion: manifest.activeVersion,
            runtimeVersion: manifest.runtimeVersion,
            manifestVersion: manifest.manifestVersion,
            bundlePath: manifest.activeBundlePath,
            installTimeMillis: manifest.installTimeMillis,
            state: manifest.state
        )
    }

    /// Removes abandoned tmp downloads, failed extraction leftovers, and anything in `downloads/`
    /// and `cache/` older than the retention window — mirrors the "obsolete" bundles Android
    /// clears, which is *only* scratch space: Android's `BundleStorage` keeps exactly one
    /// `current/` and one `rollback/` and never has more than those two real generations on disk,
    /// so cleanup never has to choose among multiple obsolete OTA versions. `current/` (active),
    /// `rollback/` (required for automatic rollback) and the app's embedded bundle are never
    /// touched here.
    func cleanupObsolete() {
        for scratch in [storage.tmpDir, storage.downloadsDir, storage.cacheDir] {
            storage.deleteRecursively(scratch)
            try? FileManager.default.createDirectory(at: scratch, withIntermediateDirectories: true)
        }
    }
}
